### PR TITLE
Add TextInput performPrivateCommand to Flutter Engine

### PR DESCRIPTION
**Pull request by guolinaileen**
## Description

Add performPrivateCommand support inside TextInputChannel

## Related Issues

This PR is part of introducing sendAppPrivateCommand and onAppPrivateCommand APIs.

## Tests

I added the following tests:
testPerformPrivateCommand_dataIsNull
testPerformPrivateCommand_dataIsByteArray
testPerformPrivateCommand_dataIsByte
testPerformPrivateCommand_dataIsCharArray
testPerformPrivateCommand_dataIsChar
testPerformPrivateCommand_dataIsCharSequenceArray
testPerformPrivateCommand_dataIsCharSequence
testPerformPrivateCommand_dataIsFloat
testPerformPrivateCommand_dataIsFloatArray

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [contributor guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [C++, Objective-C, Java style guides] for the engine.
- [x] I read the [tree hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation.
- [x] All existing and new tests are passing.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[contributor guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[tree hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by gaaclarke** [comment]
Hey @guolinaileen, can you run the formatter and add a link to an issue in the description?  It is unclear why you are adding this logic.  Thanks!

File: shell/platform/android/io/flutter/embedding/engine/systemchannels/TextInputChannel.java [hunk 1]
```diff
@@ -262,6 +264,38 @@ public void unspecifiedAction(int inputClientId) {
...
+  public void performPrivateCommand(int inputClientId, String action, Bundle data) {
+    HashMap<Object, Object> json = new HashMap<>();
+    json.put("action", action);
+    if (data != null) {
+      HashMap<String, Object> dataMap = new HashMap<>();
+      Set<String> keySet = data.keySet();
+      for (String key : keySet) {
+        Object value = data.get(key);
+        if (value instanceof byte[]) {
+          dataMap.put(key, data.getByteArray(key));
```
Comment (guolinaileen) [inline shell/platform/android/io/flutter/embedding/engine/systemchannels/TextInputChannel.java:276]: Only the ByteArray type is required for the Crowdsource Gboard project, but supporting additional types to be compatible with the types in Bundle.

**Comment by guolinaileen** [comment]
@goderbauer @justinmc @xster  Please take a look.

**Comment by guolinaileen** [comment]
> Hey @guolinaileen, can you run the formatter and add a link to an issue in the description? It is unclear why you are adding this logic. Thanks!

Any idea how to add @goderbauer @justinmc @xster  as the reviewers? They have the context of what I am doing. And how to solve this error: recipe infra failure: Infra Failure: Step('lint test') (retcode: None)?

File: shell/platform/android/io/flutter/embedding/engine/systemchannels/TextInputChannel.java (hunk 1 above)
Comment (justinmc) [reply shell/platform/android/io/flutter/embedding/engine/systemchannels/TextInputChannel.java:280]: Thanks for putting in the work to support all of those and test them all.

File: shell/platform/android/io/flutter/embedding/engine/systemchannels/TextInputChannel.java [hunk 2]
```diff
@@ -266,6 +268,38 @@ public void unspecifiedAction(int inputClientId) {
...
 
+  public void performPrivateCommand(int inputClientId, String action, Bundle data) {
+    HashMap<Object, Object> json = new HashMap<>();
+    json.put("action", action);
+    if (data != null) {
+      HashMap<String, Object> dataMap = new HashMap<>();
+      Set<String> keySet = data.keySet();
+      for (String key : keySet) {
+        Object value = data.get(key);
+        if (value instanceof byte[]) {
```
Comment (xster) [inline shell/platform/android/io/flutter/embedding/engine/systemchannels/TextInputChannel.java:279]: Didn't read the framework part yet. How would the framework figure out what type to decode into given the serialized form? Also some of this is already done via the JSONMethodCodec / JSONUtil classes. You may be able to reuse some parts.

Comment (guolinaileen) [reply shell/platform/android/io/flutter/embedding/engine/systemchannels/TextInputChannel.java:279]: The command is a client app specific command between client app and the input method. They made the agreement what command and data format that the client app can send and the input method can handle. If the command isn't supported by the input method, nothing happens. If the data getting back from the input method is not the format that the client wants, the data is not valid. In Crowdsource Gboard project the workflow like this:
From Crowdsource (client app), it sends app private command to Gboard (input method) like this: 
{
  ‘action’: 'com.google.android.apps.village.boond.GET_INPUT_CONTEXT',
  ‘data’: ‘’
}

From Gboard, it performs the app private command to Crowdsource like this:
{
  ‘action’: 'com.google.android.apps.village.boond.GET_INPUT_CONTEXT',
  ‘data’: {
    ‘input_context’: xxxxxxxxxxxxx(byte array)
  }
}

Let me explore JSONMethodCodec / JSONUtil a little bit and adjust the implementation here.

Comment (guolinaileen) [reply shell/platform/android/io/flutter/embedding/engine/systemchannels/TextInputChannel.java:279]: @xster  How to use JSONMethodCodec / JSONUtil in here? 
JSONMethodCodec is used for encoding and decoding ByteBuffer and it is not required in here. 
JSONUtil#wrap() can be used for wrapping into JSONObject, but in here we are trying to get the data type from Bundle data. In order to get the data with bundle#get<DataType>(), it has to know the data type.

Comment (xster) [reply shell/platform/android/io/flutter/embedding/engine/systemchannels/TextInputChannel.java:279]: Did you mean that for all users of performPrivateCommand, the "schema" for the data is always expected to be tightly mapped to the action?

Comment (guolinaileen) [reply shell/platform/android/io/flutter/embedding/engine/systemchannels/TextInputChannel.java:279]: Right. The "schema" of the data is mapped to the action.

Comment (xster) [reply shell/platform/android/io/flutter/embedding/engine/systemchannels/TextInputChannel.java:279]: Understood. LGTM

**Comment by justinmc** [comment]
The failure looked like an infrastructure flake to me.  I'm retrying it.

**Comment by justinmc** [comment]
No luck. @guolinaileen can you try merging master into this branch again?  Hopefully that will make it go green.

**Pull request merged**